### PR TITLE
[Perf][Attention] DeepSeek-V4.1-Flash offline throughput on 8 x B200: 24.60 s to 22.84 s per 1000 requests at TP=8, mostly from sparse-attention candidate-block kernels sized by the 1M-token limit

### PR DESCRIPTION
**Pull request by TarzanZhao**
## Purpose

`vllm bench throughput` on DeepSeek-V4.1-Flash at TP=8 spends part of every step on sparse-attention bookkeeping sized by the model's 1,048,576-token `max_model_len` instead of by the batch, and part of some steps on a DeepGEMM compile that lands inside a serving step. This PR sizes the indexer buffer by the batch, rewrites the candidate-block kernels and their top-k to walk each request's own context, and moves the output-projection compiles into engine start-up. On the one 8 x B200 node I measured, at TP=8, the 1000-request run goes from 24.60 s to 22.84 s, **1.08x**, against a stall-free baseline run; against a baseline run where one of those compiles lands in the timed window, 27.8 to 28.4 s, it is **1.22 to 1.24x**. I have not measured other GPUs, other TP sizes or online serving.

1. **Size the decode indexer logits by the batch's `max_seq_len`, not `max_model_len`.** `fp8_fp4_paged_mqa_logits` allocated a `[requests, 1,048,576]` fp32 buffer per indexer layer — 3.5 GB at 830 requests, four calls per step — and the allocator freed cached segments whenever the row count changed, a 39 ms gap in a 291 ms prefill step on a GPU that is 92% full by design. The buffer is now sized by the longest context in the batch (rounded up to 128), a host int the top-k already carried. Under CUDA graph capture that value is the worst case already, so captured shapes do not change.
2. **Candidate-block kernels walk each row's own range.** The three Triton kernels that select and apply the candidate blocks sized their work by the buffer width, and under full CUDA graphs that width is the 1M-token worst case, so every decode step wrote `-inf` into a million columns, zeroed 131k flags and scored 131k blocks for each request: 5.8 ms of a 35 ms decode step at 830 requests. They now loop over each row's own `[start, end)`; the two that tiled the width across a second grid dimension take a static `(rows, 8)` grid instead, so the launch shape stays fixed for graph replay while the work follows the context length. The selection is bit-identical to what the old kernels produced.
3. **Exact per-row top-k instead of `torch.topk` over the buffer width.** `torch.topk` ran over the full `[rows, 131k]` block-score buffer every step, about 1.2 ms at 830 requests, followed by a kernel that turned `-inf` picks into `-1`. One Triton kernel per row replaces both: a row with at most 2048 blocks (16k tokens of context) keeps all of them, a longer row runs an exact four-pass radix select over its own scores. The selected blocks come out in a different order than `torch.topk` gives, and the only consumer builds a membership table from them. Equivalence to `torch.topk`, including 131k-block rows and ties, was checked separately; that check is not shipped as a test.
4. **Bucket the o-projection token count for eager prefill steps and compile its DeepGEMM kernels at start-up.** DeepGEMM picks the configuration of the fp8 o-projection einsum from the token count and compiles a new `sm100_fp8_gemm_1d1d` kernel for each configuration it has not seen, about 3.4 s, in the serving step that first presents that count; `jit_monitor` covers Triton and CuTeDSL only, so the compile leaves no log line, and `VLLM_DEEP_GEMM_WARMUP=full` does not reach this einsum. Prefill steps above 1024 tokens now pad the o-projection input to a multiple of 1024 (zero rows, sliced off after `wo_b`, since GEMM rows are independent), and a warm-up runs each bucket once plus every token count from 1 to 1024, so the compiles happen during engine init. The costs are the padded rows and about 22 s of extra engine init the first time a DeepGEMM cache sees this model; CUDA graph shapes are untouched.

The last commit only satisfies `ruff check` and `ruff format`, which the four code commits had broken. It rewraps lines and sorts one import; the one kernel whose shape it changes (a nested `if` merged with `&`, since the operands are Triton scalars) was checked against the previous head on a B200 over eight buffer shapes, including the 131k-block production width and tie-heavy scores, and the selected blocks and masked logits are bit-identical.

## Test Plan

- `pytest tests/v1/worker` on this branch and on the unmodified base: the same environment-dependent subset fails on both (pydantic fixture errors in `test_gpu_model_runner.py`, the hybrid and Mamba cases, the ubatch slicing tests), and no failure is specific to the branch. I did not record the pass counts. No unit test is added for the new top-k kernel.
- End to end, `vllm bench throughput` on `deepseek-ai/DeepSeek-V4.1-Flash` with the ShareGPT dataset, interleaved between the two arms; the full command and the setup are under Details.
- The correctness comparison below runs from the verify branch with `PROBE=1` and the same command otherwise.

AI assistance, as `docs/contributing/README.md` asks: the code and this description were written with Claude Code, and each commit carries a `Co-Authored-By` trailer alongside its DCO sign-off.

## Test Result

Measured on one node of 8 x B200, tensor parallel 8. The metric is the benchmark's own total tokens per second over the 1000 timed requests: 444,318 tokens, the same count in every run. The branch has not been rebased since it was cut, and Mergify reports it as conflicting with the current `main`, so every number here is against its base `988d9b6777`.

| | baseline (`988d9b6777`) | this PR | change |
|---|---|---|---|
| 1000 requests, stall-free baseline runs | 24.59, 24.60 s | 22.84, 22.84 s | **1.08x** (19,452 / 18,067) |
| total tokens/s, the same runs | 18,072, 18,063 | 19,450, 19,453 | +7.7% |
| 1000 requests, baseline runs where a first-seen o-projection shape compiled inside the timed window (three of the eight timed base runs on this node) | 27.84 and 28.35 s with one compile, 30.74 s with two | no o-projection compile in the window in any run | **1.22 to 1.24x** against a one-compile run |
| peak GPU memory per rank | the same in both arms by construction: the KV cache fills to `gpu_memory_utilization` 0.92 and both arms report 421,885 KV blocks per rank; `nvidia-smi` peaked at 170,828 MiB over the branch run | | |
| engine init, not part of the metric | 111 to 114 s | 112 s with a warm DeepGEMM cache, 190 s on the first start from a new checkout path, of which the o-projection warm-up is 22.3 s for 27 kernels | |

Per commit, cumulative in branch order, single runs on another node of the same kind; the unmodified base on that node ran the same 24.6 s stall-free.

| commit | change | measured effect, 1000 requests |
|---|---|---|
| `8d946e4eef` | indexer logits sized by the batch | 24.5 s net of one compile that landed in the window (27.93 s as measured); the 16,384-token prefill steps went from 221 to 231 ms to a flat 220 to 222 ms |
| `bd45a8fd95` | static-grid candidate-block kernels | 23.37 s; decode steps at 257 to 1024 requests 23.1 ms median, from 25.6 ms |
| `e1debf8bcf` | exact per-row top-k | 22.92 s; the same decode bucket 22.2 ms median |
| `e1b3344fae` | o-projection buckets and warm-up above 1024 tokens | 22.89 s, and 22.95 s with no stall from an empty DeepGEMM cache, where the previous head had three stalls and 26.19 s |
| `b1158fa3ac` | warm-up for 1 to 1024 tokens | 22.92 s |

### Correctness

I ran the unmodified base and this branch on the same prompts with the same setup and compared 291 recorded values per run: on each of the 8 tensor-parallel ranks, the fp32 mean and max of |x| of the model output hidden states and of the logits, the sum of the argmax token ids and the batch shape for the first five real steps; plus the KV cache block count per rank and the prompt and generated token totals of the run. The tolerances were set before the runs: exact on the token totals (226,786 prompt, 217,532 generated), on the KV block count (421,885 per rank), on the argmax sums and on the batch shapes; 0.1% on the float reductions.

All 291 of 291 are identical, with no difference at rounding level either; the 0.1% is slack I declared in advance for a rounding change, and none occurred. The reference itself is stable: four runs of the unmodified code recorded the same 291 values, and the baseline record taken on this node is identical on all 291 to the record taken on the other node.

The recording code is not in this PR: it lives on the `perf/deepseek-v41-throughput-verify` branch of my fork, which is this head plus one commit that adds the calls, and the baseline record comes from `988d9b6777` plus that same commit.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

<details>
<summary>Details: hardware, model, full command, rejected settings, traces</summary>

**Hardware and environment.** One node, 8 x NVIDIA B200 (sm_100, 183 GB each), driver 580.126.20, CUDA 13.0 runtime. Python 3.12.14, torch 2.13.0+cu132, vLLM `main` at `988d9b6777` as an editable install with the precompiled cu130 kernel wheel of that commit (`VLLM_USE_PRECOMPILED=1`, no source build), flashinfer-python 0.6.18.post1 with its cubin and cu130 JIT-cache packages, triton 3.7.1, nvidia-nccl-cu13 2.29.7, transformers 5.17.0, the vendored DeepGEMM. Nothing is installed on top of that, and nothing is tuned on top of the engine defaults. Weights sit on a shared mount under `HF_HOME`; the ShareGPT file is staged to node-local disk before each run.

**Model and job.** `deepseek-ai/DeepSeek-V4.1-Flash` as published, with DeepSeek Sparse Attention at a 512-token indexer top-k and 2048 candidate blocks of 8 tokens, and `max_model_len` 1,048,576. Tensor parallel 8, no speculative decoding, engine defaults elsewhere (`max_num_batched_tokens` 16384, `max_num_seqs` 1024, `gpu_memory_utilization` 0.92, CUDA graphs FULL_AND_PIECEWISE with capture sizes up to 1024, async scheduling on by default for the `mp` executor at this commit). Each run is one process: weights 47 to 51 s, engine init 111 to 114 s, 200 warm-up prompts at seed 43 in about 10 s, then the 1000 ShareGPT prompts of seed 42 in about 25 s. That timed window is 17 to 18 chunked prefill steps of 16,384 tokens and 1646 decode steps, the batch shrinking from 927 requests to 1; 757 of the decode steps carry 1 to 4 requests, because the ShareGPT output lengths have a long tail.

**Measurement.** Both arms run the command below with the same persistent `VLLM_CACHE_ROOT` (the DeepGEMM and FlashInfer autotune caches) and `TRITON_CACHE_DIR`, the same node-local copy of the dataset, 1000 prompts at seed 42 and 200 warm-ups. Runs are interleaved baseline, branch, baseline, branch, one process each, with all eight GPUs at 0 MiB and no other job on the node before each run. Run to run, the baseline spans 0.05% (18,063 to 18,072 tok/s) and the branch 0.01% (19,450 to 19,453); the five stall-free base runs on this node span 24.56 to 24.65 s. A base run counts as stalled when the modification time of a DeepGEMM cache entry falls inside the timed window, which the gap in the benchmark's progress bar confirms. The Triton kernel `CombineTopkSwaIndicesKernel` compiles at the first timed step of every run of both arms, so it cancels out.

**Install and data** (once): `VLLM_USE_PRECOMPILED=1 uv pip install --editable . --torch-backend=auto` in the checkout, plus `flashinfer-jit-cache` for cu130 as in `docker/Dockerfile`, with `CUDA_HOME` pointing at a CUDA 13 toolkit and `nvcc` on `PATH` for the DeepGEMM and FlashInfer JIT; `ShareGPT_V4.3_unfiltered_cleaned_split.json` copied to node-local disk.

**The measured job**, both arms. The baseline arm is a checkout of `988d9b6777`, the branch arm a checkout of this head; every flag is the benchmark's, with `--num-warmups 200` added so that first-request compiles precede the timed window:

```
export HF_HOME=<hf cache> HF_HUB_OFFLINE=1 CUDA_HOME=/usr/local/cuda
export PATH=$CUDA_HOME/bin:$PATH LD_LIBRARY_PATH=$CUDA_HOME/lib64:$LD_LIBRARY_PATH
export VLLM_CACHE_ROOT=<persistent dir>      # deep_gemm/ and flashinfer_autotune_cache/ reused across runs
export TRITON_CACHE_DIR=<persistent dir>
export TMPDIR=<node-local dir> CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
vllm bench throughput --model deepseek-ai/DeepSeek-V4.1-Flash --tensor-parallel-size 8 \
  --backend vllm --dataset-name sharegpt \
  --dataset-path <ShareGPT_V4.3_unfiltered_cleaned_split.json, staged to node-local disk> \
  --num-prompts 1000 --seed 42 --num-warmups 200 --output-json <out>.json
```

There is no launcher config: `vllm bench throughput` starts the engine in-process with the multiprocess executor and 8 workers. Timing is the benchmark's own `Throughput:` line and its `elapsed_time`.

**Correctness run.** On `perf/deepseek-v41-throughput-verify` (this head plus the recording commit), the same command with `PROBE=1 PROBE_OUT=<dir>/branch/probe.json` writes the recorded values, one file per rank plus one for the totals; `988d9b6777` with the same commit cherry-picked gives `<dir>/base`. `probe compare <dir>/base <dir>/branch` prints every value against its tolerance. The recorder is [probe](https://github.com/TarzanZhao/probe), installed in the same environment and inert unless `PROBE=1`.

**Settings tried on the unmodified code and rejected.** One change at a time: `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True` (24.56 s against 24.64 s plain, inside the 0.35% run-to-run band, and engine init 134.6 s instead of 112 s), `--numa-bind` (24.65 s, the same as the plain run), `NCCL_PROTO=Simple` for the 168 MB prefill all-reduces (23.53 against 23.37 s on an earlier head). `--enable-expert-parallel` ran the job in 23.27 s (19,097 tok/s, 5.9% faster) but changes what the model computes: 144 of the 291 recorded values differ from the run without it at identical batch shapes, with the rank-0 logits abs-mean off by 0.7 to 8.5% over the first five steps, the argmax sums different on 3 of 5 steps on every rank, and 472,039 KV blocks per rank instead of 421,885. Whether that is an EP-path bug for this model at this commit or a legitimately different computation I did not determine; either way it is in neither arm. Not applicable at this commit, read in the code rather than run: `VLLM_DEEP_GEMM_WARMUP=full`, which warms `fp8_gemm_nt` linears and grouped MoE GEMMs and never touches the V4 o-projection einsum, and a torch.compile level, since the model is not decorated with `support_torch_compile` and vLLM forces `CompilationMode.NONE` for it. Left alone because they change what is served: `--max-model-len`, `--max-num-seqs`, `--max-num-batched-tokens`, `--gpu-memory-utilization`, `--kv-cache-dtype`, `--enforce-eager`, `--no-enable-prefix-caching`.

**Traces** (torch.profiler on the unmodified code, no stacks, all 8 ranks, through `--profile --profiler-config.profiler=torch`; window A is 22 steps across the prefill-to-decode transition, window B is 4 single-request tail steps; open in https://ui.perfetto.dev):

- baseline, window A: https://drive.google.com/open?id=1dmqbyK73UNuy0Xn4WYP6v_DjRRn8JQuc
- baseline, window B: https://drive.google.com/open?id=1557OZe5n53f4BPJdamlT_9fgPgiNdA96
- the branch was not profiled with torch.profiler; its per-step numbers in the tables come from a CUDA-event timer that is not part of this PR.

</details>

### CODE REVIEW

**Comment by TarzanZhao** [comment]
I have split the candidate-block kernel change out into #57459, on its own branch: one commit, two files, re-measured and re-verified against today's main rather than this PR's base. Moving this one back to draft while that is reviewed.
